src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/PreferMethodReferenceCheck.java

```diff
@@ -0,0 +1,465 @@
+// License as published by the Free Software Foundation; either
+// version 2.1 of the License, or (at your option) any later version.
+//
+// This library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+// Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public
+// License along with this library; if not, write to the Free Software
+// Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
+////////////////////////////////////////////////////////////////////////////////
+
+package com.puppycrawl.tools.checkstyle.checks.coding;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Optional;
+import java.util.Set;
+
+import com.puppycrawl.tools.checkstyle.StatelessCheck;
+import com.puppycrawl.tools.checkstyle.api.AbstractCheck;
+import com.puppycrawl.tools.checkstyle.api.DetailAST;
+import com.puppycrawl.tools.checkstyle.api.TokenTypes;
+import com.puppycrawl.tools.checkstyle.utils.TokenUtil;
+
+/**
+ * <p>
+ * Checks that lambda usage can be replaced with method reference.
+ * </p>
+ * <p>
+ * Rationale: method/constructor references are commonly agreed to be, most of the time,
+ * more compact and readable than using lambdas, and are therefore preferred.
+ * Some lambdas are not semantically equal to method references, e.g. if method invocation
+ * is performed on some expression. Such cases are not detected by default configuration,
+ * additional property {@code detectForExpression} should be set to {@code true} to enable
+ * detection for expressions.
+ * See <a href="https://docs.oracle.com/javase/tutorial/java/javaOO/methodreferences.html">link</a>
+ * for more info.
+ * </p>
+ * <ul>
+ * <li>
+ * Property {@code detectForExpression} -
+ * Enable detection for cases when method in lambda is invoked on expression.
+ * Type is {@code boolean}.
+ * Default value is {@code false}.
+ * </li>
+ * </ul>
+ * <p>
+ * To configure the check:
+ * </p>
+ * <pre>
+ * &lt;module name="PreferMethodReference"/&gt;
+ * </pre>
+ * <p>Example:</p>
+ * <pre>
+ * class Test {
+ *     BiConsumer&lt;Long, Long&gt; c = (a, b) -&gt; sum(a,b); // violation
+ *     BiFunction&lt;Object, Object, Boolean&gt; d = (a, b) -&gt; a.equals(b); // violation
+ *     Consumer&lt;Object&gt; e = arg -&gt; this.voidMethod(arg); // violation
+ *     Consumer&lt;Object&gt; f = arg -&gt; arg.voidMethod(); // violation
+ *     Function&lt;String, String&gt; g = arg -&gt; "myLiteral".concat(arg); // violation
+ *     Supplier&lt;Object&gt; s = () -&gt; new Object(); // violation
+ *     Function&lt;Integer, Object&gt; t = arg -&gt; new int[arg]; // violation
+ *     Function&lt;Long, Long&gt; exp = arg -&gt; {return a.longValue();}; // violation
```

Comment:
please add for each line compliant form.

Reply:
done

Reply:
no done.
Expected:
```
Function&lt;Long, Long&gt; exp = arg -&gt; {return a.longValue();}; // violation
Function&lt;Long, Long&gt; exp = ..................................................; // ok
```

Reply:
`// ok` comment is used when something passed check and caused no violations, therefore I used `// corrected version` instead. Do you think it still should be ok comment?

Reply:
I expect such line to have no violations, so "// ok" should be fine, but if you want you can use "// correct"

Reply:
done